components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,381 @@
+
+        playing_effect_promise
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dom-gamepadhapticactuator-reset>
+    fn Reset(&self) -> Rc<Promise> {
+        let in_realm_proof = AlreadyInRealm::assert();
+        let promise = Promise::new_in_current_realm(InRealm::Already(&in_realm_proof));
+
+        let document = self.global().as_window().Document();
+        if !document.is_fully_active() {
+            promise.reject_error(Error::InvalidState);
+            return promise;
+        }
+
+        self.sequence_id.set(self.sequence_id.get().wrapping_add(1));
+
+        if self.playing_effect_promise.borrow().is_some() {
+            let trusted_promise = TrustedPromise::new(
+                self.playing_effect_promise
+                    .borrow()
+                    .clone()
+                    .expect("Promise is null!"),
+            );
+            *self.playing_effect_promise.borrow_mut() = None;
+            let _ = self.global().gamepad_task_source().queue(
+                task!(preempt_promise: move || {
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("preempted");
+                    promise.resolve_native(&message);
+                }),
+                &self.global(),
+            );
+        }
+
+        assert!(self.playing_effect_promise.borrow().is_none());
+
+        *self.playing_effect_promise.borrow_mut() = Some(promise.clone());
+
+        self.reset_sequence_id.set(self.sequence_id.get());
+
+        let context = Trusted::new(self);
+        let (effect_stop_sender, effect_stop_receiver) =
+            ipc::channel().expect("ipc channel failure");
+        let (task_source, canceller) = (
+            self.global().gamepad_task_source(),
+            self.global().task_canceller(TaskSourceName::Gamepad),
+        );
+        let listener = HapticEffectStopListener {
+            canceller,
+            task_source,
+            context,
+        };
+
+        ROUTER.add_route(
+            effect_stop_receiver.to_opaque(),
+            Box::new(move |message| {
+                let msg = message.to::<bool>();
+                match msg {
+                    Ok(msg) => listener.handle(msg),
+                    Err(err) => warn!("Error receiving a GamepadMsg: {:?}", err),
+                }
+            }),
+        );
+
+        let router_id = GamepadRouterId::new();
+        let _ = self
+            .global()
+            .script_to_constellation_chan()
+            .send(ScriptMsg::NewGamepadRouter(
+                router_id,
+                effect_stop_sender.clone(),
+                self.global().origin().immutable().clone(),
+            ));
+
+        let event =
+            EmbedderMsg::StopGamepadHapticEffect(self.gamepad_index as usize, effect_stop_sender);
+        self.global().as_window().send_to_embedder(event);
+
+        self.playing_effect_promise.borrow().clone().unwrap()
```

Comment:
this should be `resetResultPromise`.

Reply:
`resetResultPromise` was removed in the draft version of the spec https://pr-preview.s3.amazonaws.com/w3c/gamepad/201/471b047...432bf05.html#dom-gamepadhapticactuator-reset

Reply:
Ok, thanks for the update!